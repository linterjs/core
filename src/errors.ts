import { oneLine } from "common-tags";

export class LinterProviderLoadError extends Error {
  message = oneLine`
    There was a problem importing the linter provider module: 
    "${this.linterProviderModuleName}".
  `;

  constructor(private linterProviderModuleName: string) {
    super();
  }
}

export class ModuleNotLinterProviderError extends Error {
  message = oneLine`
    The imported module is not a linter provider: 
    "${this.linterProviderModuleName}"
  `;

  constructor(private linterProviderModuleName: string) {
    super();
  }
}

export class NoLinterProvidersError extends Error {
  message = oneLine`
    No linter providers found.
    Please install at least one linter provider package.
  `;
}
