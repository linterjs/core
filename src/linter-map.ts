export interface LinterAdapterFormatInput {
  filePath?: string;
  text: string;
}

export interface LinterAdapterFormatOutput {}

export interface LinterAdapterLintInput {
  filePath?: string;
  text: string;
}

export interface LinterAdapterLintOutput {}

// XXX: LinterAdapters decide if they do anything with text baed on filePath and text
// Do we need to forward more info to LinterAdapters?
export interface LinterAdapter {
  format({
    filePath,
    text
  }: LinterAdapterFormatInput): LinterAdapterFormatOutput;
  lint({ filePath, text }: LinterAdapterLintInput): LinterAdapterLintOutput;
}

// TODO: Support config, configPath, modulePath, etc
export type LinterFactory = () => LinterAdapter;

export const linterMap = new Map<string, LinterFactory>();

export function registerLinter(
  name: string,
  linterFactory: LinterFactory
): void {
  if (linterMap.has(name)) {
    throw new Error(`Linter with name: ${name} already registered.`);
  }

  linterMap.set(name, linterFactory);
}

// XXX: Add more functionality, enable/disable linters?
