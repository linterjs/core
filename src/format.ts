import { linterMap } from "./linter-map";

export function format({ filePath, text }: { filePath: string, text: string }) {
  // TODO: Format all the things!
}
