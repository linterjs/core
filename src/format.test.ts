import { NoLintersError } from "./errors";
import { format } from "./format";
import { registerLinter } from "./linter-map";

describe("Format", () => {
  const filePath = "test.js";
  const text = 'const foo = "bar"';

  test("No registered linters", async () => {
    try {
      await format({ text });
    } catch (error) {
      expect(error).toBeInstanceOf(NoLintersError);
    }
  });

  test("text only", async () => {
    const { linter } = require("@linter/eslint");
    const args = { text };
    const result = await format(args);
    expect(result).toMatchSnapshot();
    expect(linter.format).toHaveBeenCalledTimes(1);
    expect(linter.format).toHaveBeenCalledWith(args);
  });

  test("text and filePath", async () => {
    const { linter } = require("@linter/eslint");
    const args = { filePath, text };
    const result = await format(args);
    expect(result).toMatchSnapshot();
    expect(linter.format).toHaveBeenCalledTimes(2);
    expect(linter.format).toHaveBeenCalledWith(args);
  });

  describe("with Prettier registered", () => {
    test("Text only", async () => {
      const { linter } = require("@linter/prettier");
      const args = { text };
      const result = await format(args);
      expect(result).toMatchSnapshot();
      expect(linter.format).toHaveBeenCalledTimes(1);
      expect(linter.format).toHaveBeenCalledWith(args);
    });

    test("text and filePath", async () => {
      const { linter } = require("@linter/prettier");
      const args = { text, filePath };
      const result = await format(args);
      expect(result).toMatchSnapshot();
      expect(linter.format).toHaveBeenCalledTimes(2);
      expect(linter.format).toHaveBeenCalledWith(args);
    });
  });
});
