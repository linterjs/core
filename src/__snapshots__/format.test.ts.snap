// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Format text and filePath 1`] = `
Object {
  "errorCount": 0,
  "filePath": "test.js",
  "messages": Array [],
  "output": "const foo = \\"bar\\"",
  "warningCount": 0,
}
`;

exports[`Format text only 1`] = `
Object {
  "errorCount": 0,
  "messages": Array [],
  "output": "const foo = \\"bar\\"",
  "warningCount": 0,
}
`;

exports[`Format with Prettier registered Text only 1`] = `
Object {
  "errorCount": 0,
  "messages": Array [],
  "output": "prettier:const foo = \\"bar\\"",
  "warningCount": 0,
}
`;

exports[`Format with Prettier registered text and filePath 1`] = `
Object {
  "errorCount": 0,
  "filePath": "test.js",
  "messages": Array [],
  "output": "prettier:const foo = \\"bar\\"",
  "warningCount": 0,
}
`;
